File: VB/E2442_UnboundDataFromDetail/Form1.vb
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Data
Imports System.Windows.Forms
Imports DevExpress.Data
Imports DevExpress.Data.Filtering.Helpers
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid

Namespace E2442_UnboundDataFromDetail
	Partial Public Class Form1
		Inherits Form

		Private grid As GridControl = Nothing
		Private catView As GridView = Nothing
		Private prodView As GridView = Nothing

		Public Sub New()
			InitializeComponent()
		End Sub

		Private dataSet1 As New DataSet()
		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
			grid = New GridControl()
			grid.Parent = Me
			grid.Dock = DockStyle.Fill
			grid.ForceInitialize()

			Dim categories = DataHelper.CreateCategoriesTable()
			categories.TableName = "Categories"
			dataSet1.Tables.Add(categories)
			dataSet1.Tables.Add(DataHelper.CreateProductsTable())
			Dim keyColumn As DataColumn = dataSet1.Tables(0).Columns("CategoryID")
			Dim foreignKeyColumn As DataColumn = dataSet1.Tables(1).Columns("CategoryID")
			dataSet1.Relations.Add("CategoriesProducts", keyColumn, foreignKeyColumn)

			grid.MainView.Dispose()

			catView = New GridView(grid)
			prodView = New GridView(grid)

			grid.MainView = catView
			grid.LevelTree.Nodes.Add("CategoriesProducts", prodView)

			grid.DataSource = dataSet1.Tables(0)
			AddHandler catView.CustomUnboundColumnData, AddressOf OnUnboundColumnData

			Dim colIndex As Integer = catView.Columns.Add(New GridColumn())
			catView.Columns(colIndex).FieldName = "Field"
			catView.Columns(colIndex).UnboundType = UnboundColumnType.Decimal
			catView.Columns(colIndex).Caption = "Detail price sum"
			catView.Columns(colIndex).Visible = True
		End Sub

		Private Sub OnUnboundColumnData(ByVal sender As Object, ByVal e As CustomColumnDataEventArgs)
			If Not e.IsGetData Then
				Return
			End If

			Dim pdc As PropertyDescriptorCollection = DirectCast(dataSet1.Tables(1).DefaultView, ITypedList).GetItemProperties(Nothing)
			Dim evaluator As New ExpressionEvaluator(pdc, prodView.ActiveFilterCriteria)

			Dim controllerRow As Integer = catView.DataController.GetControllerRow(e.ListSourceRowIndex)
			If Not catView.DataController.IsValidControllerRowHandle(controllerRow) Then
				Return
			End If

			Dim detailList As IList = catView.DataController.GetDetailList(controllerRow, 0)
			If detailList.Count = 0 Then
				Return
			End If

			Dim sum As Decimal = 0
			For i As Integer = 0 To detailList.Count - 1
				If evaluator.Fit(DirectCast(detailList(i), DataRowView)) Then
					sum += Convert.ToDecimal(DirectCast(detailList(i), DataRowView)("UnitPrice"))
				End If
			Next i

			e.Value = sum
		End Sub
	End Class
End Namespace

File: CS/E2442_UnboundDataFromDetail/Form1.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Data;
using System.Windows.Forms;
using DevExpress.Data;
using DevExpress.Data.Filtering.Helpers;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;

namespace E2442_UnboundDataFromDetail
{
	public partial class Form1 : Form
	{
		GridControl grid = null;
		GridView catView = null;
		GridView prodView = null;

		public Form1()
		{
			InitializeComponent();
		}

        DataSet dataSet1 = new DataSet();
        private void Form1_Load(object sender, EventArgs e)
		{
			grid = new GridControl();
			grid.Parent = this;
			grid.Dock = DockStyle.Fill;
			grid.ForceInitialize();

            var categories = DataHelper.CreateCategoriesTable();
            categories.TableName = "Categories";
            dataSet1.Tables.Add(categories);
            dataSet1.Tables.Add(DataHelper.CreateProductsTable());
            DataColumn keyColumn = dataSet1.Tables[0].Columns["CategoryID"];
            DataColumn foreignKeyColumn = dataSet1.Tables[1].Columns["CategoryID"];
            dataSet1.Relations.Add("CategoriesProducts", keyColumn, foreignKeyColumn);

            grid.MainView.Dispose();

			catView = new GridView(grid);
			prodView = new GridView(grid);

			grid.MainView = catView;
			grid.LevelTree.Nodes.Add("CategoriesProducts", prodView);

            grid.DataSource = dataSet1.Tables[0];
			catView.CustomUnboundColumnData += new CustomColumnDataEventHandler(OnUnboundColumnData);

			int colIndex = catView.Columns.Add(new GridColumn());
			catView.Columns[colIndex].FieldName = "Field";
			catView.Columns[colIndex].UnboundType = UnboundColumnType.Decimal;
			catView.Columns[colIndex].Caption = "Detail price sum";
			catView.Columns[colIndex].Visible = true;
		}

        private void OnUnboundColumnData(object sender, CustomColumnDataEventArgs e)
		{
			if ( !e.IsGetData )
				return;

			PropertyDescriptorCollection pdc = ((ITypedList)dataSet1.Tables[1].DefaultView).GetItemProperties(null);
			ExpressionEvaluator evaluator = new ExpressionEvaluator(pdc, prodView.ActiveFilterCriteria);

			int controllerRow = catView.DataController.GetControllerRow(e.ListSourceRowIndex);
			if ( !catView.DataController.IsValidControllerRowHandle(controllerRow) )
				return;

			IList detailList = catView.DataController.GetDetailList(controllerRow, 0);
			if ( detailList.Count == 0 )
				return;

			decimal sum = 0;
			for ( int i = 0; i < detailList.Count; i++ )
				if ( evaluator.Fit((DataRowView)detailList[i]) )
					sum += Convert.ToDecimal(((DataRowView)detailList[i])["UnitPrice"]);

			e.Value = sum;
		}
	}
}